### PR TITLE
Add new rule: `require-input-label`

### PR DESCRIPTION
**Pull request by MelSumner**
This adds a new rule that provides a minimum check to see if an input element has met the following conditions: 

1. it's wrapped in a label element
2. it has one of these three attributes: id, aria-label, or aria-labelledby

### CODE REVIEW

File: lib/rules/require-input-label.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+              message: ERROR_MESSAGE,
+              line: node.loc && node.loc.start.line,
+              column: node.loc && node.loc.start.column,
+              source: this.sourceForNode(node),
+            });
+          }
+        },
+        exit(node) {
+          if (node.tag === 'label') {
+            this._parentStack.pop();
```
Comment (Turbo87) [inline lib/rules/require-input-label.js]: why are we only popping from the parent stack if it's a label?

**Comment by Turbo87** [comment]
I couldn't help myself:

Bildschirmfoto 2020-02-07 um 10 27 29

I think this is what you initially wanted to do?

File: docs/rule/require-input-label.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+This rule **forbids** the following:
+
+```hbs
+<div><input /></div>
+```
+
+This rule **allows** the following:
+
+```hbs
+<label><input /></label>
```
Comment (bmish) [inline docs/rule/require-input-label.md]: Shouldn't this example actually have some label text?

File: test/unit/rules/require-input-label-test.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    '<label>LabelText<Input /></label>',
+    '<label><Input />LabelText</label>',
+    '<label><div><input /></div></label>', // technically okay, hopefully no one does this though
+    '<input id="probablyHasLabel" />', // it's likely to have an associated label if it has an id attribute
+    '<input aria-label={{labelText}} />',
+    '<input aria-labelledby="someIdValue" />',
+  ],
+
+  bad: [
+    {
```
Comment (bmish) [inline test/unit/rules/require-input-label-test.js]: Can you add a lone `<input />` element as a test case?

File: test/unit/rules/require-input-label-test.js [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+const generateRuleTests = require('../../helpers/rule-test-harness');
+const ERROR_MESSAGE = require('../../../lib/rules/require-input-label').ERROR_MESSAGE;
+
+generateRuleTests({
+  name: 'require-input-label',
+
+  config: true,
+
+  good: [
```
Comment (bmish) [inline test/unit/rules/require-input-label-test.js]: Can you add a test case of `<div>` to make sure unrelated elements are ignored?

File: lib/helpers/iterate-parents.js [hunk 5]
```diff
@@ -0,0 +1,8 @@
+function* iterateParents(path) {
```
Comment (rwjblue) [inline lib/helpers/iterate-parents.js]: I think we can remove this now that https://github.com/glimmerjs/glimmer-vm/pull/1024 landed and is released.

File: docs/rule/require-input-label.md [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+```
+
+```hbs
+<input ...attributes />
+```
+
+### Migration
+
+* the simplest fix is to add an aria-label to the input element. 
+* the most robust fix is to add an associated label element and use CSS to visually hide the label while still making it available to screen readers.
```
Comment (rwjblue) [inline docs/rule/require-input-label.md]: I think this suggestion is a bit off. It is assuming that no label previously existed _on purpose_ (e.g. that the label display was unwanted). I think that is much less common than "I just forgot", and the simplest / most semantic solution is to add the label (leaving it displayed)...

File: lib/helpers/iterate-parents.js (hunk 5 above)
Comment (rwjblue) [reply lib/helpers/iterate-parents.js]: This utility needs to be removed, and update the internals of the rule to use `path.parents` directly

File: lib/rules/require-input-label.js [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+        }
+
+        // OR:
+        // Parental validation (descriptive elements)
+        let parents = Array.from(path.parents());
+        if (
+          parents.some(parent => parent.node.type === 'ElementNode' && parent.node.tag === 'label')
+        ) {
+          return;
+        }
```
Comment (rwjblue) [inline lib/rules/require-input-label.js]: Can you extract this to a new `AstNodeInfo.hasParent(path, filterFunction)`, which would be used here like:

```js
let hasLabelParent = AstNodeInfo.hasParent(path, parent => parent.node.type === 'ElementNode' && parent.node.tag === 'label');
if (hasLabelParent) {
  return;
}
```

File: test/unit/rules/require-input-label-test.js [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    '<label><Input />LabelText</label>',
+    '<label><div><input /></div></label>', // technically okay, hopefully no one does this though
+    '<input id="probablyHasLabel" />', // it's likely to have an associated label if it has an id attribute
+    '<input aria-label={{labelText}} />',
+    '<input aria-labelledby="someIdValue" />',
+    '<div></div>',
+    '<input ...attributes/>', // we are unable to correctly determine if this has a label or not, so we have to allow it
+  ],
+
+  bad: [
```
Comment (rwjblue) [inline test/unit/rules/require-input-label-test.js]: Please add additional "bad" examples for each of the following:

A label wrapping an input, but without label text:

```hbs
<label><input></label>
```

Usage of `{{input}}`:

```hbs
{{input}}
```

```hbs
<div>{{input}}</div>
```

Usage of `<Input/>`:

```hbs
<Input />
```

File: test/unit/rules/require-input-label-test.js [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+const generateRuleTests = require('../../helpers/rule-test-harness');
+const ERROR_MESSAGE = require('../../../lib/rules/require-input-label').ERROR_MESSAGE;
+
+generateRuleTests({
+  name: 'require-input-label',
+
+  config: true,
+
+  good: [
```
Comment (rwjblue) [inline test/unit/rules/require-input-label-test.js]: Please add additional "good" tests for:

```hbs
<Input ...attributes>
<Input id="foo" />
{{input id=someThing}}
<label>Text here<Input /></label>
```

File: docs/rule/require-input-label.md (hunk 6 above)
Comment (MelSumner) [reply docs/rule/require-input-label.md]: Please feel free to suggest a specific edit.

File: docs/rule/require-input-label.md [hunk 10]
```diff
@@ -0,0 +1,60 @@
+## require-valid-label
+
+Users with assistive technology need `<input>` elements to have associated labels. It's so essential for users, that failure to provide an associated label for an `<input>` element will cause failure of three different WCAG success criteria: [1.3.1, Info and Relationships](https://www.w3.org/WAI/WCAG21/Understanding/info-and-relationships.html), [3.3.2, Labels or Instructions](https://www.w3.org/WAI/WCAG21/Understanding/labels-or-instructions.html), and [4.1.2, Name, Role, Value](https://www.w3.org/WAI/WCAG21/Understanding/name-role-value.html). It is also associated with common failure [#68: Failure of Success Criterion 4.1.2 due to a user interface control not having a programmatically determined name](https://www.w3.org/WAI/WCAG21/Techniques/failures/F68).
```
Comment (MelSumner) [inline docs/rule/require-input-label.md]:
```suggestion
Users with assistive technology need `<input>` elements to have associated labels. It's so essential for users that failure to provide an associated label for an `<input>` element will cause failure of **three** different WCAG success criteria: [1.3.1, Info and Relationships](https://www.w3.org/WAI/WCAG21/Understanding/info-and-relationships.html), [3.3.2, Labels or Instructions](https://www.w3.org/WAI/WCAG21/Understanding/labels-or-instructions.html), and [4.1.2, Name, Role, Value](https://www.w3.org/WAI/WCAG21/Understanding/name-role-value.html). It is also associated with common failure [#68: Failure of Success Criterion 4.1.2 due to a user interface control not having a programmatically determined name](https://www.w3.org/WAI/WCAG21/Techniques/failures/F68).
```

File: lib/rules/require-input-label.js [hunk 11]
```diff
@@ -0,0 +1,75 @@
+'use strict';
+
+const Rule = require('./base');
+const AstNodeInfo = require('../helpers/ast-node-info');
+// const iterateParents = require('../helpers/iterate-parents');
```
Comment (rwjblue) [inline lib/rules/require-input-label.js]: we should remove this comment

Comment (MelSumner) [inline lib/rules/require-input-label.js]:
```suggestion
```

**Pull request merged**